File: 1444-number-of-ways-of-cutting-a-pizza/1444-number-of-ways-of-cutting-a-pizza.java
class Solution {
    public int ways(String[] pizza, int k) {
        int mod = 1000000007;
        int m = pizza.length, n = pizza[0].length();
        int[][] grid = new int[m+1][n+1];
        int[][][] dp = new int[m+1][n+1][k];
        for (int i = m - 1; i > -1; i--) {
            for (int j = n - 1; j > -1; j--) {
                grid[i][j] = pizza[i].charAt(j) == 'A' ? 1 : 0;
                grid[i][j] += grid[i+1][j] + grid[i][j+1] - grid[i+1][j+1];
                dp[i][j][0] = grid[i][j] == 0 ? 0 : 1;
            }
        }
        for (int p = 1; p < k; p++) {
            for (int q = m - 1; q > -1; q--) {
                for (int r = n - 1; r > -1; r--) {
                    for (int s = q + 1; s < m; s++) {
                        int apples = grid[q][r] - grid[s][r];
                        if (apples > 0) dp[q][r][p] += dp[s][r][p-1];
                        dp[q][r][p] %= mod;
                    }
                    for (int s = r + 1; s < n; s++) {
                        int apples = grid[q][r] - grid[q][s];
                        if (apples > 0) dp[q][r][p] += dp[q][s][p-1];
                        dp[q][r][p] %= mod;
                    }
                }
            }
        }
        return dp[0][0][k-1];
        
    }
}

File: 1444-number-of-ways-of-cutting-a-pizza/1444-number-of-ways-of-cutting-a-pizza.py
class Solution:
    def ways(self, pizza: List[str], k: int) -> int:
        m, n = len(pizza), len(pizza[0])
        grid = [[0] * (n+1) for i in range(m+1)]
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, - 1, -1):
                grid[i][j] = 1 if pizza[i][j] == 'A' else 0
                grid[i][j] += grid[i+1][j] + grid[i][j+1] - grid[i+1][j+1]
        
        @cache
        def res(i = 0, j = 0, k = k):
            mod = 10**9+7
            if k == 1:
                #print(i, j, k, 1 if grid[i][j] > 0 else 0)
                return 1 if grid[i][j] > 0 else 0
            elif i == m or j == n:
                return 0
            else:
                ret = 0
                for l in range(i+1, m):
                    top = grid[i][j] - grid[l][j]
                    if top > 0:
                        ret += res(l, j, k - 1)
                for l in range(j+1, n):
                    left = grid[i][j] - grid[i][l]
                    if left > 0:
                        ret += res(i, l, k - 1)
                #print(i, j, k, ret)
                return ret%mod
        
        return res()
                